phoenix-core/src/main/java/org/apache/phoenix/schema/SchemaExtractionTool.java

```diff
@@ -0,0 +1,488 @@
+        String pSchemaName = table.getSchemaName().getString();
+        String pTableName = table.getTableName().getString();
+        String baseTableName = table.getParentTableName().getString();
+        String baseTableFullName = SchemaUtil.getQualifiedTableName(pSchemaName, baseTableName);
+        PTable baseTable = getPTable(baseTableFullName);
+        String columnInfoString = getColumnInfoStringForView(table, baseTable);
+
+        String whereClause = table.getViewStatement();
+        if(whereClause != null) {
+            whereClause = whereClause.substring(whereClause.indexOf("WHERE"));
+        }
+        return generateCreateViewDDL(columnInfoString, baseTableFullName, whereClause == null ? "" : " "+whereClause, pSchemaName, pTableName);
+    }
+
+    private String generateCreateViewDDL(String columnInfoString, String baseTableFullName, String whereClause, String pSchemaName, String pTableName) {
+        String viewFullName = SchemaUtil.getQualifiedTableName(pSchemaName, pTableName);
+        StringBuilder outputBuilder = new StringBuilder(String.format(CREATE_VIEW, viewFullName, columnInfoString, baseTableFullName, whereClause));
+        return outputBuilder.toString();
+    }
+
+    public String extractCreateTableDDL(PTable table) throws IOException, SQLException {
+
+        String pSchemaName = table.getSchemaName().getString();
+        String pTableName = table.getTableName().getString();
+
+        ConnectionQueryServices cqsi = getCQSIObject();
+        HTableDescriptor htd = getHTableDescriptor(cqsi, table);
+        HColumnDescriptor hcd = htd.getFamily(SchemaUtil.getEmptyColumnFamily(table));
+
+        populateDefaultProperties(table);
+        setPTableProperties(table);
+        setHTableProperties(htd);
+        setHColumnFamilyProperties(hcd);
+
+        String columnInfoString = getColumnInfoStringForTable(table);
+        String propertiesString = convertPropertiesToString();
+
+        return generateTableDDLString(columnInfoString, propertiesString, pSchemaName, pTableName);
+    }
+    private String generateTableDDLString(String columnInfoString,String propertiesString,String pSchemaName,String pTableName) {
+        String pTableFullName = SchemaUtil.getQualifiedTableName(pSchemaName, pTableName);
+        StringBuilder outputBuilder = new StringBuilder(String.format(CREATE_TABLE, pTableFullName));
+        outputBuilder.append(columnInfoString).append(" ").append(propertiesString);
+        return outputBuilder.toString();
+    }
+
+    private void populateDefaultProperties(PTable table) {
+        Map<String, String> propsMap = HColumnDescriptor.getDefaultValues();
+        for (Map.Entry<String, String> entry : propsMap.entrySet()) {
+            String key = entry.getKey();
+            String value = entry.getValue();
+            defaultProps.put(key, value);
+            if(key.equalsIgnoreCase(BLOOMFILTER) || key.equalsIgnoreCase(COMPRESSION)) {
```

Comment:
This is just to override the values that are specifically defaulted in Phoenix. Otherwise, we use HColumnDescriptor.getDefaultValues() to get all version-specific default values for hbase.